File: artifacts/chaincode/identity/createIdentity.go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// Identity struct représente une identité dans le système
type Identity struct {
	DID       string `json:"did"`       // Identifiant décentralisé
	PublicKey string `json:"publicKey"` // Clé publique associée à l'identité
	Metadata  string `json:"metadata"`  // Métadonnées supplémentaires (ex. méthodes d'authentification)
	X509Cert  string `json:"x509Cert"`  // Certificat X.509 associé à l'identité
}

// SmartContract définit la structure du contrat intelligent
type SmartContract struct {
	contractapi.Contract
}

// CreateIdentity crée une nouvelle identité avec un DID et un certificat X.509
func (s *SmartContract) CreateIdentity(ctx contractapi.TransactionContextInterface, did string, publicKey string, metadata string, x509Cert string) error {
	// Créer une nouvelle instance d'Identity
	identity := Identity{
		DID:       did,
		PublicKey: publicKey,
		Metadata:  metadata,
		X509Cert:  x509Cert,
	}

	// Convertir l'identité en JSON
	identityJSON, err := json.Marshal(identity)
	if err != nil {
		return fmt.Errorf("échec de la conversion de l'identité en JSON: %v", err)
	}

	// Enregistrer l'identité dans l'état mondial (ledger)
	err = ctx.GetStub().PutState(did, identityJSON)
	if err != nil {
		return fmt.Errorf("échec de l'enregistrement de l'identité dans l'état mondial: %v", err)
	}

	return nil
}

// ReadIdentity récupère une identité à partir du ledger en utilisant le DID
func (s *SmartContract) ReadIdentity(ctx contractapi.TransactionContextInterface, did string) (*Identity, error) {
	// Lire l'identité depuis l'état mondial
	identityJSON, err := ctx.GetStub().GetState(did)
	if err != nil {
		return nil, fmt.Errorf("échec de la lecture de l'identité: %v", err)
	}
	if identityJSON == nil {
		return nil, fmt.Errorf("l'identité %s n'existe pas", did)
	}

	// Convertir le JSON en struct Identity
	var identity Identity
	err = json.Unmarshal(identityJSON, &identity)
	if err != nil {
		return nil, fmt.Errorf("échec de la conversion du JSON en identité: %v", err)
	}

	return &identity, nil
}

// UpdateIdentity met à jour le certificat X.509 pour une identité donnée
func (s *SmartContract) UpdateIdentity(ctx contractapi.TransactionContextInterface, did string, x509Cert string) error {
	// Récupérer l'identité existante
	identity, err := s.ReadIdentity(ctx, did)
	if err != nil {
		return err
	}

	// Mettre à jour le certificat X.509
	identity.X509Cert = x509Cert
	identityJSON, err := json.Marshal(identity)
	if err != nil {
		return fmt.Errorf("échec de la conversion de l'identité mise à jour en JSON: %v", err)
	}

	// Enregistrer l'identité mise à jour dans l'état mondial
	return ctx.GetStub().PutState(did, identityJSON)
}

// main initialise le chaincode
func main() {
	// Créer une instance du chaincode
	chaincode, err := contractapi.NewChaincode(new(SmartContract))
	if err != nil {
		log.Panicf("Erreur lors de la création du chaincode Identity: %v", err)
	}

	// Démarrer le chaincode
	if err := chaincode.Start(); err != nil {
		log.Panicf("Erreur lors du démarrage du chaincode Identity: %v", err)
	}
}

File: artifacts/chaincode/identity/createIdentity.js
'use strict';

const { Contract } = require('fabric-contract-api');

class IdentityContract extends Contract {

    // Crée un nouveau document DID
    async createDID(ctx, did, publicKey, metadata, x509Cert) {
        const didDocument = {
            did,
            publicKey,
            metadata,
            x509Cert,
            createdAt: new Date().toISOString(),
        };

        const didJSON = JSON.stringify(didDocument);

        // Stocke le document DID dans l'état de la chaîne
        await ctx.stub.putState(did, Buffer.from(didJSON));
    }

    // Récupère un document DID par DID
    async readDID(ctx, did) {
        const didBytes = await ctx.stub.getState(did);
        if (!didBytes || didBytes.length === 0) {
            throw new Error(`Document DID ${did} non trouvé`);
        }

        return didBytes.toString(); // Retourne les informations du document DID sous forme de chaîne
    }

    // Met à jour le certificat X.509 dans le document DID
    async updateDID(ctx, did, x509Cert) {
        const didBytes = await ctx.stub.getState(did);
        if (!didBytes || didBytes.length === 0) {
            throw new Error(`Document DID ${did} non trouvé`);
        }

        const didDocument = JSON.parse(didBytes.toString());
        didDocument.x509Cert = x509Cert; // Met à jour le certificat X.509

        const didJSON = JSON.stringify(didDocument);
        await ctx.stub.putState(did, Buffer.from(didJSON));
    }
}

// Instanciation du chaincode
module.exports = IdentityContract;